### ISSUE TITLE
Small ternary issue && ... || ...

### ISSUE THREAD

**Issue by drcmda** [prettier/prettier#2867]
Not that it made much sense in that case when ? ... : ... would do the same, but i accidentally stumbled over it while rewriting a previous ternary expression.

**Prettier 1.7.0**
[Playground link](https://prettier.io/playground/#%7B%22content%22%3A%22let%20test%20%3D%20%7B%5Cn%20%20%20%20plugins%3A%20%5B%5Cn%20%20%20%20%20%20%20%20new%20WebpackNotifierPlugin()%2C%5Cn%20%20%20%20%20%20%20%20...((env.prod%20%26%26%20%5B%5Cn%20%20%20%20%20%20%20%20%20%20%20%20new%20webpack.DefinePlugin(%7B%20'process.env'%3A%20%7B%20NODE_ENV%3A%20JSON.stringify('production')%20%7D%20%7D)%2C%5Cn%20%20%20%20%20%20%20%20%20%20%20%20new%20webpack.LoaderOptionsPlugin(%7B%20minimize%3A%20true%2C%20debug%3A%20false%20%7D)%2C%5Cn%20%20%20%20%20%20%20%20%20%20%20%20new%20webpack.optimize.UglifyJsPlugin(%7B%20compress%3A%20%7B%20warnings%3A%20false%20%7D%2C%20mangle%3A%20true%20%7D)%2C%5Cn%20%20%20%20%20%20%20%20%5D)%20%7C%7C%20%5B%5D)%2C%5Cn%20%20%20%20%5D%2C%5Cn%7D%22%2C%22options%22%3A%7B%22ast%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22doc%22%3Afalse%2C%22jsxBracketSameLine%22%3Atrue%2C%22output2%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22printWidth%22%3A120%2C%22semi%22%3Afalse%2C%22singleQuote%22%3Atrue%2C%22tabWidth%22%3A4%2C%22trailingComma%22%3A%22all%22%2C%22useTabs%22%3Afalse%7D%7D)
```sh
--jsx-bracket-same-line
--print-width 120
--no-semi
--single-quote
--tab-width 4
--trailing-comma all
```

**Input:**
```jsx
let test = {
    plugins: [
        new WebpackNotifierPlugin(),
        ...((env.prod && [
            new webpack.DefinePlugin({ 'process.env': { NODE_ENV: JSON.stringify('production') } }),
            new webpack.LoaderOptionsPlugin({ minimize: true, debug: false }),
            new webpack.optimize.UglifyJsPlugin({ compress: { warnings: false }, mangle: true }),
        ]) || []),
    ],
}
```

**Output:**
```jsx
let test = {
    plugins: [
        new WebpackNotifierPlugin(),
        ...((env.prod && [
            new webpack.DefinePlugin({ 'process.env': { NODE_ENV: JSON.stringify('production') } }),
            new webpack.LoaderOptionsPlugin({ minimize: true, debug: false }),
            new webpack.optimize.UglifyJsPlugin({ compress: { warnings: false }, mangle: true }),
        ]) ||
            []),
    ],
}

```

**Expected behavior:**

**Comment by duailibe** [comment]
Indeed. Why do we only inline array and objects when they're not empty? i.e.

```
const foo =
  bar &&
  baz &&
  [];

const foo = bar &&
    baz && [0];
```

**Comment by duailibe** [comment]
Okay now I get why it's like this.. goes back to #1342 but I'm not sure the fix was correct (i.e. the example above)

**Comment by AHH0623** [comment]
good
